### PR TITLE
Issue a deprecation warning when compiling with Visual Studio 2017

### PR DESCRIPTION
**Pull request by bernhardmgruber**

### CODE REVIEW

File: cmake/CCCLBuildCompilerTargets.cmake [hunk 1]
```diff
@@ -24,6 +24,7 @@ option(CCCL_ENABLE_EXCEPTIONS "Enable exceptions within CCCL libraries." ON)
 option(CCCL_ENABLE_RTTI "Enable RTTI within CCCL libraries." ON)
 option(CCCL_ENABLE_WERROR "Treat warnings as errors for CCCL targets." ON)
 option(CCCL_SUPPRESS_ICC_DEPRECATION_WARNING "Suppress Intel Compiler deprecation warnings" OFF)
+option(CCCL_SUPPRESS_MSVC2017_DEPRECATION_WARNING "Suppress Visual Studio 2017 deprecation warnings" OFF)
```
Comment (miscco) [inline cmake/CCCLBuildCompilerTargets.cmake:27]: Question: Should we combine them into just `CCCL_SUPPRESS_COMPILER_DEPRECATION_WARNING`

We might collect a bunch of those

**Review by miscco — APPROVED** [review]

File: libcudacxx/include/cuda/std/__cccl/compiler.h [hunk 2]
```diff
@@ -56,6 +56,10 @@
 #elif defined(_MSC_VER)
 #  define _CCCL_COMPILER_MSVC     _CCCL_COMPILER_MAKE_VERSION(_MSC_VER / 100, _MSC_VER % 100)
 #  define _CCCL_COMPILER_MSVC2017 (_CCCL_COMPILER_MSVC < _CCCL_COMPILER_MAKE_VERSION(19, 20))
+#  if _CCCL_COMPILER_MSVC2017 && !defined(CCCL_SUPPRESS_MSVC2017_DEPRECATION_WARNING)
+#    warning \
+      "Support for the Visual Studio 2017 (MSC_VER < 1920) is deprecated and will eventually be removed. Define CCCL_SUPPRESS_MSVC2017_DEPRECATION_WARNING to suppress this warning"
+#  endif // CCCL_SUPPRESS_ICC_DEPRECATION_WARNING
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__cccl/compiler.h:62]: `warning` does not work, we need  
```suggestion
#  if _CCCL_COMPILER_MSVC2017 && !defined(CCCL_SUPPRESS_MSVC2017_DEPRECATION_WARNING)
#    pragma message(\
      "Support for the Visual Studio 2017 (MSC_VER < 1920) is deprecated and will eventually be removed. Define CCCL_SUPPRESS_MSVC2017_DEPRECATION_WARNING to suppress this warning")
#  endif // CCCL_SUPPRESS_ICC_DEPRECATION_WARNING
```

**Review by miscco — APPROVED** [review]

File: libcudacxx/include/cuda/std/__cccl/compiler.h (hunk 2 above)
Comment (bernhardmgruber) [reply libcudacxx/include/cuda/std/__cccl/compiler.h:62]: Right. I copied this from the ICC deprecation PR. `#warning` did not work on MSVC until C++23 standardized it.

**Comment by bernhardmgruber** [comment]
pre-commit.ci autofix

File: cmake/CCCLBuildCompilerTargets.cmake (hunk 1 above)
Comment (bernhardmgruber) [reply cmake/CCCLBuildCompilerTargets.cmake:27]: #2996

**Pull request merged**